db/redshift.js

```diff
@@ -0,0 +1,220 @@
+    if (error) {
+      rollbar.error('Error in transaction: ', error)
+      redshiftClient.query('ROLLBACK', (error) => {
+        if (error) {
+          rollbar.error('Error rolling back: ', error)
+        }
+      })
+    }
+    return !!error
+  }
+
+  const commit = async () => {
+    return new Promise((resolve, reject) => {
+      redshiftClient.query('COMMIT').then(() => {
+        resolve()
+      })
+      .catch((error) => {
+        throw new Error('Error committing transaction: ' + error)
+      })
+    })
+  }
+
+  const createStagingTable = async (stagingTable, permanentTable) => {
+    return new Promise((resolve, reject) => {
+      const createQuery = `CREATE TEMP TABLE ${stagingTable} (LIKE scale_of_belief.${permanentTable})`
+      redshiftClient.query(createQuery).then(() => {
+        resolve()
+      })
+      .catch((error) => {
+        if (shouldAbort(error)) {
+          throw new Error(error)
+        }
+      })
+    })
+  }
+
+  const copyDataFromS3 = async (stagingTable, bucketFolder) => {
+    const copyFromS3 =
+      `COPY ${stagingTable}
+      FROM '${process.env.REDSHIFT_S3_BUCKET}/${bucketFolder}/${bucketDate}.csv'
+      IAM_ROLE '${process.env.REDSHIFT_IAM_ROLE}'
+      CSV IGNOREHEADER 1 BLANKSASNULL EMPTYASNULL TRUNCATECOLUMNS COMPUPDATE OFF STATUPDATE OFF`
+
+    return new Promise((resolve, reject) => {
+      redshiftClient.query(copyFromS3).then(() => {
+        resolve()
+      })
+      .catch((error) => {
+        if (shouldAbort(error)) {
+          throw new Error(error)
+        }
+      })
+    })
+  }
+
+  const mergeDataIntoPermanentTable = async (stagingTable, permanentTable, primaryKey) => {
+    const deleteQuery =
+      `DELETE FROM scale_of_belief.${permanentTable}
+      USING ${stagingTable}
+      WHERE scale_of_belief.${permanentTable}.${primaryKey} = ${stagingTable}.${primaryKey}`
+
+    const insertQuery =
+      `INSERT INTO scale_of_belief.${permanentTable}
+      SELECT * FROM ${stagingTable}`
+
+    return new Promise((resolve, reject) => {
+      redshiftClient.query(deleteQuery).then(() => {
+        redshiftClient.query(insertQuery).then(() => {
+          resolve()
+        })
+        .catch((error) => {
+          if (shouldAbort(error)) {
+            throw new Error(error)
+          }
+        })
+      })
+      .catch((error) => {
+        if (shouldAbort(error)) {
+          throw new Error(error)
+        }
+      })
+    })
+  }
+
+  const dropTable = async (tableName) => {
+    return new Promise((resolve, reject) => {
+      redshiftClient.query('DROP TABLE ' + tableName).then(() => {
+        resolve()
+      })
+      .catch((error) => {
+        if (shouldAbort(error)) {
+          throw new Error(error)
+        }
+      })
+    })
+  }
+
+  const moveData = async (tableName, stagingTable, primaryKey) => {
+    const now = Date.now()
+    const lowerThreshold = new Date(now - (10 * 60 * 1000)).toISOString() // 10 minutes ago
```

Comment:
Should we store a last run success time (maybe in redis) and use that to determine `lowerThreshold`? Since the lambda runs every 5 minutes, if it failed more than twice, would it miss data?

Reply:
Yeah, it probably would miss data. I don't know much about redis, I've never used it. We can try adding a mechanic like that.